src/box/upgrade.c

```diff
@@ -0,0 +1,419 @@
+	 * returns array i.e. result should be if the form of "array of arrays":
+	 * MP_ARRAY MP_ARRAY..
+	 */
+	assert(mp_typeof(**new_data_begin) == MP_ARRAY);
+	uint32_t arr_sz = mp_decode_array(new_data_begin);
+	if (arr_sz != 1 || mp_typeof(**new_data_begin) != MP_ARRAY) {
+		diag_set(ClientError, ER_UPGRADE, space_name(space),
+			"type of return value is expected to be array");
+		return -1;
+	}
+	return 0;
+}
+
+/**
+ * Function provides verification of three conditions:
+ * 1. If new tuple satisfy new space format;
+ * 2. If new tuple has the same PK;
+ * 3. If new tuple remains unique (see comment in the body below).
+ */
+static int
+tuple_upgrade_check_tuple(struct space *space, struct tuple *old_tuple,
+			  const char *new_tuple, const char *new_tuple_end)
+{
+	assert(space_is_memtx(space));
+	struct tuple_format *new_format = space->def->opts.upgrade->format;
+	assert(new_format != NULL);
+	/*
+	 * tuple_new() checks that converted tuple satisfies new format in
+	 * terms of field types.
+	 */
+	struct tuple *tuple = tuple_new(new_format, new_tuple, new_tuple_end);
+	if (tuple == NULL)
+		return -1;
+	int rc = 0;
+	/* Check that PK isn't changed. */
+	struct key_def *pk_key_def = space->index[0]->def->key_def;
+	/* Upgrade operation is not supported for multikey indexes. */
+	assert(!pk_key_def->is_multikey);
+	if (tuple_compare(old_tuple, HINT_NONE, tuple, HINT_NONE,
+			  pk_key_def) != 0) {
+		diag_set(ClientError, ER_CANT_UPDATE_PRIMARY_KEY,
+			 space->index[0]->def->name, space_name(space));
+		rc = -1;
+		goto finish;
+	}
+	/* Now check that unique constraints will be met. We consider that
+	 * new key is OK in terms of uniqueness in case after upgrade it
+	 * doesn't conflict with OLD keys (except the case when key remains
+	 * unchanged). Imagine space contains three tuples and unique index
+	 * covering them:
+	 * [1] -upgrade-> [4] -- it is OK: 4 is unique (among old keys)
+	 * [2] -upgrade-> [3] -- it is not OK: key 3 already exists
+	 * [3] -upgrade-> [x] -- OK in case x >= 4 or x <= 0
+	 */
+	for (uint32_t i = 1; i < space->index_count; ++i) {
+		struct index *idx = space->index[i];
+		if (! idx->def->opts.is_unique)
+			continue;
+		struct key_def *key_def = idx->def->key_def;
+		uint32_t new_key_size = 0;
+		const char *new_key =
+			tuple_extract_key(tuple, key_def, MULTIKEY_NONE,
+					  &new_key_size);
+		if (new_key == NULL) {
+			rc = -1;
+			goto finish;
+		}
+		uint32_t part_count = mp_decode_array(&new_key);
+		assert(part_count == key_def->part_count);
+		struct tuple *found = 0;
+		if (index_get(idx, new_key, part_count, &found) != 0) {
+			rc = -1;
+			goto finish;
+		}
+		if (found != NULL) {
+			if (tuple_compare(found, HINT_NONE, tuple, HINT_NONE,
+					  pk_key_def) != 0) {
+				diag_set(ClientError, ER_TUPLE_FOUND,
+					 idx->def->name,
+					 space_name(space),
+					 tuple_str(found),
+					 tuple_str(tuple));
+				rc = -1;
+				goto finish;
+			}
+		}
+	}
```

Comment:
Don't get your comment:(  Could you please elaborate on it?

Reply:
AFAICS you could implement the "test" mode by running exactly the same code as in "upgrade" mode with one difference: call `txn_rollback` instead of `txn_commit` in the end.